hpx/util/checkpoint.hpp

```diff
@@ -0,0 +1,516 @@
+        int64_t length;
+        ist.read(reinterpret_cast<char*>(&length), sizeof(int64_t));
+        ckp.data.resize(length);
+        // Read in the next checkpoint
+        ist.read(ckp.data.data(), length);
+        return ist;
+    }
+
+    //Function object for save_checkpoint
+    namespace detail {
+        struct save_funct_obj
+        {
+            template <typename... Ts>
+            checkpoint operator()(checkpoint&& c, Ts&&... ts) const
+            {
+                //Create serialization archive from checkpoint data member
+                hpx::serialization::output_archive ar(c.data);
+                //Serialize data
+                int const sequencer[] = {//Trick to expand the variable pack
+                    (ar << ts,
+                        0)...};      //Takes advantage of the comma operator
+                (void) sequencer;    // Suppress unused param. warnings
+                return c;
+            }
+        };
+    }
+
+    ///////////////////////////////////
+    /// Save_checkpoint
+    ///
+    /// \tparam T            Containers passed to save_checkpoint to be
+    ///                      serialized and placed into a checkpoint object.
+    ///
+    /// \tparam Ts           More containers passed to save_checkpoint
+    ///                      to be serialized and placed into a
+    ///                      checkpoint object.
+    ///
+    /// \tparam U            This parameter is used to make sure that T
+    ///                      is not a launch policy or a checkpoint. This
+    ///                      forces the compiler to choose the correct overload.
+    ///
+    /// \param t             A container to restore.
+    ///
+    /// \param ts            Other containers to restore Containers
+    ///                      must be in the same order that they were
+    ///                      inserted into the checkpoint.
+    ///
+    /// Save_checkpoint takes any number of objects which a user may wish
+    /// to store and returns a future to a checkpoint object.
+    /// Additionally the function can take a policy as a first object which
+    /// changes its behavior depending on the policy passed to it. Most
+    /// notably, if a sync policy is used save_checkpoint will simply return a
+    /// checkpoint object.
+    ///
+    /// \returns Save_checkpoint returns a future to a checkpoint with one
+    ///          exception: if you pass hpx::launch::sync as the first
+    ///          argument. In this case save_checkpoint will simply return
+    ///          a checkpoint.
+
+    template <typename T, typename... Ts,
+        typename U =
+            typename std::enable_if<!hpx::traits::is_launch_policy<T>::value &&
+                !std::is_same<typename std::decay<T>::type,
+                    checkpoint>::value>::type>
+    hpx::future<checkpoint> save_checkpoint(T&& t, Ts&&... ts)
+    {
+        {
+            return hpx::dataflow(detail::save_funct_obj(),
```

Comment:
Shouldn't we use the io service pool here?

Reply:
> Shouldn't we use the io service pool here?

@sithhell what for?

Reply:
?